### PR TITLE
Write cache atomically using a tempfile in the same directory

### LINKED ISSUE #1107: Cached filesystem not concurrency-safe?
We have a program that uses fsspec to mount a cached HTTP filesystem as a FUSE mount; while this process runs in the background, another traverses the FUSE mount and inspects multiple files in parallel.  Unfortunately, the FUSE process keeps hitting errors in the following form:

```
Uncaught exception from FUSE operation open, returning errno.EINVAL.
Traceback (most recent call last):
  File "/home/dandi/cronlib/dandisets-healthstatus/venv/lib/python3.8/site-packages/fuse.py", line 734, in _wrapper
    return func(*args, **kwargs) or 0
  File "/home/dandi/cronlib/dandisets-healthstatus/venv/lib/python3.8/site-packages/fuse.py", line 834, in open
    fi.fh = self.operations('open', path.decode(self.encoding),
  File "/home/dandi/cronlib/dandisets-healthstatus/venv/lib/python3.8/site-packages/datalad_fuse/fuse_.py", line 75, in __call__
    return super(DataLadFUSE, self).__call__(op, self.root + path, *args)
  File "/home/dandi/cronlib/dandisets-healthstatus/venv/lib/python3.8/site-packages/fuse.py", line 1076, in __call__
    return getattr(self, op)(*args)
  File "/home/dandi/cronlib/dandisets-healthstatus/venv/lib/python3.8/site-packages/datalad_fuse/fuse_.py", line 203, in open
    fsspec_file = self._adapter.open(path)
  File "/home/dandi/cronlib/dandisets-healthstatus/venv/lib/python3.8/site-packages/datalad_fuse/fsspec.py", line 243, in open
    return dsap.open(relpath, mode=mode, encoding=encoding, errors=errors)
  File "/home/dandi/cronlib/dandisets-healthstatus/venv/lib/python3.8/site-packages/datalad_fuse/fsspec.py", line 169, in open
    return self.fs.open(url, mode, **kwargs)
  File "/home/dandi/cronlib/dandisets-healthstatus/venv/lib/python3.8/site-packages/fsspec/implementations/cached.py", line 406, in <lambda>
    return lambda *args, **kw: getattr(type(self), item).__get__(self)(
  File "/home/dandi/cronlib/dandisets-healthstatus/venv/lib/python3.8/site-packages/fsspec/spec.py", line 1034, in open
    f = self._open(
  File "/home/dandi/cronlib/dandisets-healthstatus/venv/lib/python3.8/site-packages/fsspec/implementations/cached.py", line 406, in <lambda>
    return lambda *args, **kw: getattr(type(self), item).__get__(self)(
  File "/home/dandi/cronlib/dandisets-healthstatus/venv/lib/python3.8/site-packages/fsspec/implementations/cached.py", line 345, in _open
    self.save_cache()
  File "/home/dandi/cronlib/dandisets-healthstatus/venv/lib/python3.8/site-packages/fsspec/implementations/cached.py", line 406, in <lambda>
    return lambda *args, **kw: getattr(type(self), item).__get__(self)(
  File "/home/dandi/cronlib/dandisets-healthstatus/venv/lib/python3.8/site-packages/fsspec/implementations/cached.py", line 161, in save_cache
    cached_files = pickle.load(f)
EOFError: Ran out of input
```

I suspect the cause is that the cached filesystem is not safe for concurrent access, with the result that multiple actions on the filesystem are causing the cache file to be read by one procedure while another writes to it.  Indeed, there seems to be no locking in the cached filesystem code.

### PR DESCRIPTION
**Pull request by jwodder**
Closes #1107.

### CODE REVIEW

File: fsspec/implementations/cached.py [hunk 1]
```diff
@@ -834,3 +833,19 @@ def hash_name(path, same_name):
     else:
         hash = hashlib.sha256(path.encode()).hexdigest()
     return hash
+
+
+@contextlib.contextmanager
+def atomic_write(path, mode="wb"):
```
Comment (yarikoptic) [inline fsspec/implementations/cached.py:839]: Looks great to me, thank you!  I  would have only appreciated a docstring briefly describing the purpose/implementation of this function.

**Comment by yarikoptic** [comment]
FWIW: Although doesn't really resolve #1107 I think this PR is the right thing to do.  WDYT @martindurant ?

**Comment by yarikoptic** [comment]
not sure why downstream testing errored out

```
Updating 'test_env' env from conda env update...
  /usr/share/miniconda/condabin/conda env update --name test_env --file ci/environment-downstream.yml
  Collecting package metadata (repodata.json): ...working... done
  Solving environment: ...working... ::endgroup::
Error: The process '/usr/share/miniconda/condabin/conda' failed with exit code null
```

and have no super-powers to restart, but I think this PR is ready for being reviewed/merged (attn @martindurant )

**Comment by martindurant** [comment]
Fixing CI in https://github.com/zarr-developers/zarr-python/pull/1237 , probably needs a few iterations yet

**Pull request merged**